go-semconv/generator.go

```diff
@@ -0,0 +1,437 @@
+
+	err = format(cfg.outputFilename)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type config struct {
+	inputPath        string
+	outputPath       string
+	outputFilename   string
+	templateFilename string
+	containerImage   string
+	specVersion      string
+}
+
+func validateConfig(cfg config) (config, error) {
+	if cfg.inputPath == "" {
+		return config{}, errors.New("input path must be provided")
+	}
+
+	if cfg.outputFilename == "" {
+		cfg.outputFilename = fmt.Sprintf("%s.go", path.Base(cfg.inputPath))
+	}
+
+	if cfg.specVersion == "" {
+		// Find the latest version of the specification and use it for generation.
+		var err error
+		cfg.specVersion, err = findLatestSpecVersion(cfg)
+		if err != nil {
+			return config{}, err
+		}
+	}
+
+	if cfg.outputPath == "" {
+		// If output path is unspecified put it under a sub-directory with a name matching
+		// the version of semantic convention under the semconv directory.
+		cfg.outputPath = path.Join("semconv", cfg.specVersion)
+	}
+
+	if !path.IsAbs(cfg.outputPath) {
+		root, err := FindRepoRoot()
+		if err != nil {
+			return config{}, err
+		}
+		cfg.outputPath = path.Join(root, cfg.outputPath)
+	}
+
+	cfg.outputFilename = path.Join(cfg.outputPath, cfg.outputFilename)
+
+	if !path.IsAbs(cfg.templateFilename) {
+		pwd, err := os.Getwd()
+		if err != nil {
+			return config{}, err
+		}
+		cfg.templateFilename = path.Join(pwd, cfg.templateFilename)
+	}
+
+	return cfg, nil
+}
+
+func render(cfg config) error {
+	tmpDir, err := os.MkdirTemp("", "otel_semconvgen")
+	if err != nil {
+		return fmt.Errorf("unable to create temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	specCheckoutPath := path.Join(tmpDir, "input")
+	err = os.Mkdir(specCheckoutPath, 0700)
+	if err != nil {
+		return fmt.Errorf("unable to create input directory: %w", err)
+	}
+
+	outputPath := path.Join(tmpDir, "output")
+	err = os.Mkdir(outputPath, 0700)
+	if err != nil {
+		return fmt.Errorf("unable to create output directory: %w", err)
+	}
+
+	// Checkout the specification repo to a temp dir. This will be the input
+	// for the generator.
+	doneFunc, err := checkoutSpecToDir(cfg, specCheckoutPath)
+	if err != nil {
+		return err
+	}
+	defer doneFunc()
+
+	err = exec.Command("cp", cfg.templateFilename, tmpDir).Run()
+	if err != nil {
+		return fmt.Errorf("unable to copy template to temp directory: %w", err)
+	}
+
+	cmd := exec.Command("docker", "run", "--rm",
+		"-v", fmt.Sprintf("%s:/data", tmpDir),
+		cfg.containerImage,
+		"--yaml-root", path.Join("/data/input/semantic_conventions/", path.Base(cfg.inputPath)),
+		"code",
+		"--template", path.Join("/data", path.Base(cfg.templateFilename)),
+		"--output", path.Join("/data/output", path.Base(cfg.outputFilename)),
+	)
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("unable to render template: %w", err)
+	}
+
+	err = os.MkdirAll(cfg.outputPath, 0700)
+	if err != nil {
+		return fmt.Errorf("unable to create output directory %s: %w", cfg.outputPath, err)
+	}
```

Comment:
nit: running this prior to the "expensive" rendering operation would allow a quick failure if this fails.